Algorithmic Methods for Data Mining
==============================


## Homework 3 - Airbnb Search Engine
### Group #17 - Giulia Scialanga, Guilherme Nicchio, Marco Minici
##### 26/11/2018

The goal of this project consists in buiding a search engine over a data base of Airbnb houses. 

The code returns the houses of the data base which matches the descriptions entered by an user query. 

User query means the sentence an user enter in a search field, for example: "A beuatiful house with beach and garden".

For this project was used the following libraries:

In [2]:
from os.path import isdir
from os import mkdir
import pandas as pd
import numpy as np
import csv
import math
from os import listdir
from os.path import isfile
from os import remove
from tqdm import tqdm
from sklearn.feature_extraction.text import TfidfVectorizer
from geopy import distance
import folium
import os
from heapq import nlargest
from nltk import SnowballStemmer
from nltk.tokenize import RegexpTokenizer
from nltk.corpus import stopwords
import re
import string
import pickle
import inflect

The functions created for the purpose of this work are in the file **airbnb_search_engine.py** located in the repository. For displaying the results of this work let's import it and assign the main class of it to a variable.

In [39]:
import airbnb_search_engine
airbnb = airbnb_search_engine.AirBnbPy()

# The Data

The data used is a csv file with the following columns.

In [40]:
data = airbnb.loadData()
data.head()

average_rate_per_night bedrooms_count         city date_of_listing  \
1                    $27              2       Humble        May 2016   
2                   $149              4  San Antonio   November 2010   
3                    $59              1      Houston    January 2017   
4                    $60              1        Bryan   February 2016   
5                    $75              2   Fort Worth   February 2017   

                                         description   latitude  longitude  \
1  Welcome to stay in private room with queen bed...  30.020138 -95.293996   
2  Stylish, fully remodeled home in upscale NW – ...  29.503068 -98.447688   
3  'River house on island close to the city' \nA ...  29.829352 -95.081549   
4  Private bedroom in a cute little home situated...  30.637304 -96.337846   
5  Welcome to our original 1920's home. We recent...  32.747097 -97.286434   

                                               title  \
1    2 Private rooms/bathroom 10min from IAH airport   
2  Unique Location! Alamo Heights - Designer Insp...   
3                          River house near the city   
4                       Private Room Close to Campus   
5                                          The Porch   

                                                 url  
1  https://www.airbnb.com/rooms/18520444?location...  
2  https://www.airbnb.com/rooms/17481455?location...  
3  https://www.airbnb.com/rooms/16926307?location...  
4  https://www.airbnb.com/rooms/11839729?location...  
5  https://www.airbnb.com/rooms/17325114?location...

The first goal on the process is to match the user query with descriptions and tittles of the houses.

## Cleaning the data

#### Cleaning process:

It was noticed that the data base has some "noise" within it, for example empty cells (NA) in location coordinates, title and description which is essential for the concept of search engine built and the houses of the respective residences won't produce any match to an entered query.

It was also noticed duplicated cells, that sometimes are system error or a house entered in the system twice by the owner, in order to avoid a duplicated outcome for the query the duplicated cells will be dropped out of the data base.

Cleaning process:

-Delete all rows with "latitude" equal to NA, in this cases when latitude is empty was noticed that also longitude column was empty, thus if a house is not locatable it is useless for a customer.

-Delete rows with **both** description and title equal to NA since it wouldn't be possible to match the user query.

-Retain all others rows.

#### 1.1.1 Load the data.csv

- read csv file with pandas csv read.

#### 1.1.2 Call cleandata()

- Gets just data which has not null in latitude column;

- Drop NA's from description and title;

- Drop all duplicated rows;

- Returns the data.


## Pre-processing the data and creating Tsv files for each house
At this point there is a data frame with duplicated removed and empty fields removed.
The next step is to go through each row of the data frame process it and write a tsv file with it.

Before diving into text and feature extraction, the first step will be further treating the data in order to obtain better features. We will achieve this by doing some of the basic _*pre-processing*_ steps on our data, for this it was used the library NLTK which has many features for natural language processing.

### Pre-processing

###### Lower case
The first pre-processing step is to transform our descriptions and titles into lower case. This avoids having multiple copies of the same words. For example, while calculating the word count, ‘House’ and ‘house’ would be taken as different words.

##### Tokenization
Tokenization refers to dividing the text into a sequence of words or sentences.

##### Removing Punctuation and Stop Words
The next step is to remove punctuation and stopwords, as it doesn’t add any extra information while treating text data. Therefore removing all instances of it will help us reduce the size of the data.

##### Stemming
Stemming refers to the removal of suffices, like “ing”, “ly”, “s”, etc. by a simple rule-based approach. For this purpose, it will be used SnowballStemmer from the NLTK library.

##### Removing non-enlish characters
When handling the data was noticed non-english texts, for example descriptions entered in chinese language can bring problematics when processing the text. In order to solve this issue was created a function to test if the description an title are indeed using english characters, therefore, the data just proceed for process and creating tsv file if the english test function returns true for the test.

### Creating TSV files
After pre processing the text we join the description and title text and replace then in the original field, proceeding to the tsv file creation. The treated data has 18259 rolls which will result in the same number of tsv files for the analysis of the search engine with the title and description properly pre-processed

# Building the Search Engine

This section consists in:

- Build the encoding for the words in the corpus. To each word is associated a unique int ID.

- Create the reverted index. To each word ID is associated the set of documents which contains that word.

- Do an interface to perform queries to the search engine.

### Building the Encoding
At this point the tsv files are created and pre-processed. The next step is to create a dictionary for the vocabulary of words present in all the documents, in this dictionary each word corresponds to a single number. This process will help analyse which documents have the respective word.

Example of what is the target at this point: {"car":1, "house":2}

This step was approached browsing through every file's title and description and adding its words into an empty dictionary. But before adding it is checked if the word isn't yet in the dictionary, when adding the word to the dictionary its key value is a counter of how many unique words were collected so far. 

Algorithm

    1.BuildingEncoding
        Create an empty dictionary to store words
        Create a counter to store the index of how many single words we have (vocabulary size)
        for each file in "listdir(directory address)"
            Goes to the title, strip its words and create a set of it to remove duplicates
            do the same for the description
            Union both sets
            For each word in the united set check if it is not dictionary already
                If it isnt add it to the dictionary as a key
                Set the counter as the value of the word in the dictionary
                Increase the counter +1

### Reverse Index
Creating a dictionary of words as keys and documents as values.

After this step can say in for each number (word) which documents have it example 

 
{

term_id_1:[document_1, document_2, document_4],

term_id_2:[document_1, document_3, document_5, document_6],

...}


In this step it is created a new dictionary which the keys are the the vocabulary numbers of the previous dictionary. Then it is checkd for each document if the title and the description has the respective word, if there is the document name is appended as a value to the respective keyword.

In [44]:
reverted_index = airbnb.RI

In [45]:
len(reverted_index)

12385

We have 12385 different words we have in all the documents.

Algorithm

    Create a new dictionary with the keys being the numbers at the size of the vocabulary
        this is our target dictionary, we want to say what documents have this word
    For each file in "listdir(directory address)"
        Goes to the title, strip its words and create a set of it to remove duplicates
        Do the same for the description
        Union both sets
        For each word in the set
            Get which number this word has in the vocabulary dictionary
            Goes to the new dictionary and append the name of the document to the key

# Conjunctive query

At this moment, we narrow our interest on the description and title of each document. It means that the first Search Engine will evaluate queries with respect to the aforementioned information.

Since we are dealing with conjunctive queries (AND), each of the returned documents should contain all the words in the query. The final output is a table of documents with title, description, city and url.

Our first goal is to get the query and process it in order to be in the same standard as the description and title.

### Query processing

After the user enter a query, which is considered a sentence which will be acquired in string format and apply the same process of Tokenization, eliminating stopwords and stemming the words, which were described in the **Pre-processing**.

For example, a query **"a beautiful house with garden and beach"** turns into a list **['beauti', 'hous', 'garden', 'beach']**

Now that we have the query properly formated its possible to go to the dictionary created in the previous section and check which documents have the words of the query.

In this step we go through each word in the query list and append the set of documents of present on the dictionary values to a new list. At the end of the preccess this new list has all sets of documents which contain the query words

As it is wanted a document that has all the words in it we do an intersection of this sets, which correspond to the group of documents that have all the words.

### Example

First let's create the term encoding and the reverted index dictionary to perform the query.

In [43]:
airbnb.buildEncoding(fileName = "term_encoding.pickle")
airbnb.createRevertedIndex(fileName = "reverted_index.pickle")

Imputing the query and performing the search.

In [18]:
result,_ = airbnb.query(str(input("Enter here your search: ")))
print(result)

Enter here your search: a beautiful house with garden and beach
Query processed:  ['beauti', 'hous', 'garden', 'beach']
Empty DataFrame
Columns: [title, description, city, url]
Index: []


Note that the query function returned a empty data frame, this means that there are no results matching this query. Let's try to simplify the query.

In [20]:
result,_ = airbnb.query(str(input("Enter here your search: ")))
result.head()

Enter here your search: a house at the beach
Query processed:  ['hous', 'beach']


title  \
4067   Bring the whole family & the dog. 5 min to beach!   
1696   Unwind & Play | Sun-Drenched Galveston Beach H...   
5839          League City, TX Rental 3 bedroom, 2 baths.   
13351   Historic Cole St House, walk to water, Cole Park   
17090                                      The COVE & RV   

                                             description            city  \
4067   Our beach house has everything you need: comfo...    Port Aransas   
1696   Welcome to Camp Lafitte, a classic Galveston b...       Galveston   
5839   1600 sq ft house in League City, TX, a great l...     League City   
13351  1600 Square foot historic house in a great nei...  Corpus Christi   
17090  PIRATES COVE and RV is a beautiful beachfront ...   Crystal Beach   

                                                     url  
4067   https://www.airbnb.com/rooms/13686358?location...  
1696   https://www.airbnb.com/rooms/16484229?location...  
5839   https://www.airbnb.com/rooms/16603529?location...  
13351  https://www.airbnb.com/rooms/14211883?location...  
17090  https://www.airbnb.com/rooms/6870793?location=...

In [21]:
len(result)

320

This query matched 320 houses.

## Conjunctive query & Ranking score

Let's improve the Search Engine, given a query, now let's get the top 5 documents related to the query.

For this, will be retrieved all the documents that contains all the words in the query, as before. But afterwards they will be sorted by their similarity with the query

To compute the simalarity will be used the tf-Idf score, and the Cosine similarity.

The TF-IDF weight function ( term frequency-inverse document frequency ) is a function used in information retrieval to measure the importance of a term with respect to a document or a collection of documents. This function increases proportionally to the number of times the term is contained in the document, but grows in an inversely proportional manner with the frequency of the term in the collection. The idea behind this behavior is to give more importance to the terms that appear in the document, but that are generally infrequent.

Cosine similarity calculates similarity by measuring the cosine of angle between two vectors. This is calculated as:

\begin{equation}
cos(\pmb x, \pmb y) = \frac {\pmb x \cdot \pmb y}{||\pmb x|| \cdot ||\pmb y||}
\end{equation}

With cosine similarity, we need to convert sentences into vectors. One way to do that is to use bag of words with either TF (term frequency) or TF-IDF (term frequency- inverse document frequency). The choice of TF or TF-IDF depends on application and is immaterial to how cosine similarity is actually performed — which just needs vectors. TF is good for text similarity in general, but TF-IDF is good for search query relevance.

### The computation flow as follow.

In order to evaluate the cosine similarity we must prepare our code to execute it efficiently.

### Build corpus for the TF-IDF

Computing the TFIDF values involve creating the corpus of our dataset. The corpus consists of a collection of documents where each document is represented by the string obtained concatenating the title and the description field of each record.

### Build the TF-IDF matrix

The TF-IDF matrix is formed and then the frequency of each word in a document is available as show the generic example below:

|    -   |This  | is   |a     |word    |in    |the   |document|
|--------|------|------|------|--------|------|------|--------|
|doc_1   |0     |0.234 |0     |0.468      |0.234     |0     |0.468        |
|doc_2   |0.234     |0.234     |0.468      |0       |0     |0.234    |0.234       |

In which each frequency of the word is normalized, having then a value between 0 and 1.

### Build  the term encoding and the reverted index

After this step a dictionary is available in which for each word there is the list of documents in which it is contained in, and its relative tf-Idf score.

**Example:**

{

**term_id_1**:[(document1, tfIdf_{term,document1}), (document2, tfIdf_{term,document2}), (document4, tfIdf_{term,document4}), ...],

**term_id_2**:[(document1, tfIdf_{term,document1}), (document3, tfIdf_{term,document3}), (document5, tfIdf_{term,document5}), (document6, tfIdf_{term,document6}), ...],

...}

**Important remark**: Before creating the corpus, tf-idf matrix, encoding and reversed index, first it is checked if it was already created previously, it will just build them if was not before.

### Perform the Cosine Similarity
First it is performed the Conjuctive query without ranking for the query inputed by the user.

Then retrieve the indexes of the docuents that match the non-ranked query.

Perform the cosine similarity for each document of the non-ranked, for this it is used the vectorize function wich allows to apply the cosine similarity for all members of the array.

### Example

In [42]:
result = airbnb.rankedQuery('a beautiful house', k =  10)

Query processed:  ['beauti', 'hous']


100%|███████████████████████████████████████████████████████████████████████████| 18036/18036 [00:51<00:00, 349.26it/s]


[LOG]: The file tfIdfMatrix.pickle doesn't exist
[LOG]: A new tfidf matrix will be built and saved in persistent memory with the name:= tfIdfMatrix.pickle
[LOG]: The file term_encodingTFIDF.pickle doesn't exist
[LOG]: A new term encoding will be built and saved in persistent memory with the name:= term_encodingTFIDF.pickle
[LOG]: The file RI2.pickle doesn't exist
[LOG]: A new inverted index will be built and saved in persistent memory with the name:= RI2.pickle


In [43]:
result

title  \
0                                        Quiet place   
1                         Beautiful, New Beach House   
2                         My house is big and quite.   
3  Beautiful home near the splash pad, park and p...   
4  Beautiful home near the splash pad, park and p...   
5                                 House on the Creek   
6  House in wonderful neighborhood & beautiful views   
7  House in wonderful neighborhood & beautiful views   
8                          1. Room in beautiful home   
9     Quite Spacious Private Room in Beautiful House   

                                         description           city  \
0                                    Beautiful house      New Caney   
1  Our house is close to restaurants and dining, ...  Crystal Beach   
2  My house is a Mexican rustic decorate. Big roo...    San Antonio   
3  The house is one house away from a park and .7...           Katy   
4  The house is one house away from a park and .7...           Katy   
5  A beautiful house on the Creek, conveniently l...          Plano   
6  My place is good for couples, solo adventurers...    San Antonio   
7  My place is good for couples, solo adventurers...    San Antonio   
8  A nice clean room in a quiet, beautiful house ...    San Antonio   
9  Our house is in a small and quite family neigh...        Houston   

                                                 url  Similarity  
0  https://www.airbnb.com/rooms/16743790?location...    0.701582  
1  https://www.airbnb.com/rooms/15068378?location...    0.455250  
2  https://www.airbnb.com/rooms/7378323?location=...    0.352277  
3  https://www.airbnb.com/rooms/19079555?location...    0.348313  
4  https://www.airbnb.com/rooms/19079555?location...    0.348313  
5  https://www.airbnb.com/rooms/17992009?location...    0.346690  
6  https://www.airbnb.com/rooms/17345573?location...    0.345514  
7  https://www.airbnb.com/rooms/17345573?location...    0.345514  
8  https://www.airbnb.com/rooms/13915253?location...    0.343841  
9  https://www.airbnb.com/rooms/18784719?location...    0.339466

# Now let's define a new score

Let's be more proactive and ask the user to input some information to improve the previous result for the entered query.

For example, would be interesting for the user choose minimal amount of rooms, a location desired for the house and the maximum price it is willing to pay.

Let's imagine that we provide a platform where the user input in the respective fields the amount of rooms, the price also the user is able to point a desired location in map.

With this is possible to compute the similarity of each house to the target inputed by the user.

**The workflow for the developed code go as follow**

For each document, it is collected the number of rooms, the price, the latitude and longitude and the similarity is a weighted mean of each individual distance, outputing a number between 1 and 0 where 0 is the maximum match.

The weight of each similarity in the calculation of the final similarity was defined through trial until the results were satisfactory.

\begin{equation}
Similarity = weight rooms \times Rooms Similarity + weight loc \times Location Similarity + weight price \times Price Similarity
\end{equation}

Where the Rooms Similarity is calculated as:

\begin{equation}
RoomsSimilarity = \frac{1}{1 + RoomsDifference^{10}}
\end{equation}

The Location Similarity was calculated using the library geopy and the function geodesic which allows to compute the precise ellipsoid distance from one point of coordinates to the other on earth.

\begin{equation}
LocationSimilarity = \frac{1}{1 + 0.1 \times GeoDistance}
\end{equation}

And the Price Similarity is calculated as:

\begin{equation}
PriceSimilarity = \frac{1}{\frac{House Price}{Price Requested}}
\end{equation}

Although if the number of rooms desired is smaller than the rooms of the house it shouldn't be a problem, thus in this cases the rooms similarity is set to 1.

The same annalogy goes for the price, if the house is cheaper than the requested price the price similarity is also set to 1.

The weights chosen for this search that fit for all solutions were 0.15 for the price, 0.25 for the rooms and 0.6 for the location. This parameters have proven to satisfy for many different custom queries.

In [ ]:
weightsFFA = [0.15, 0.25, 0.6] #[price, rooms, location]

### Further improving

Let's allow the user to change the weights give on the previous custom ranking by inputing what is more relevant for him.

The user might be very restrict with location and would not mind to pay more for example. Than the user can input the ranking by opting for price oriented or location oriented search.

Where the following weights would be applied if the user wants to have a more oriented ranking.

In [ ]:
weightsPrice = [0.7, 0.1, 0.2] #[price, rooms, location]
weightsLoc = [0.1, 0.2, 0.7] #[price, rooms, location]

## Example

In [46]:
#interface with the user
querytext = str(input('Enter your search: '))

queryinput =[]
queryinput.append(int(input("How much would you pay at maximum? ")))
queryinput.append(int(input("How many rooms do you want? ")))

location = list(map(float, input("Give us a location: ").split(", ")))
for x in location:
    queryinput.append(x)
    
orientation = int(input("What is more important for you? The possible choices are:\n\t[0] Fit-for-all\n\t[1] Price oriented\n\t[2] Location oriented \n\t Your Choice: "))

Enter your search: a beach house
How much would you pay at maximum? 300
How many rooms do you want? 3
Give us a location: 29.064693, -95.134938
What is more important for you? The possible choices are:
	[0] Fit-for-all
	[1] Price oriented
	[2] Location oriented 
	 Your Choice: 0


In [33]:
cQuery = airbnb.customQuery(query= queryinput, textQuery= querytext, k=5, weightChoice= orientation)

#### Standard, fit for both price and location

In [34]:
cQuery

ranking                                              title  \
14862        1  Salt Grass- Christmas Bay Great Fishing & Kaya...   
15733        2            Adorable Mid Century Modern Beach Cabin   
14939        3                         Charming Beach Front House   
13294        4                              Surfside paradise # 1   
17971        5                    Beautiful custom home ON beach!   

                                             description            city  \
14862  This Bay house is close to the beach and sits ...        Freeport   
15733  Newly remodeled second row beach house with ne...        Freeport   
14939  Charming &amp; colorful beach house. 3 bedroom...        Freeport   
13294  Comfortably sleeps 10 people. Open floor plan,...  Surfside Beach   
17971  Welcome to our home, Sea Dreams Beach House! W...        Freeport   

                                                     url  
14862  https://www.airbnb.com/rooms/13186000?location...  
15733  https://www.airbnb.com/rooms/14185321?location...  
14939  https://www.airbnb.com/rooms/4532740?location=...  
13294  https://www.airbnb.com/rooms/15073135?location...  
17971  https://www.airbnb.com/rooms/6179121?location=...

#### Price oriented

In [35]:
# weightChoice= 1
cQuery = airbnb.customQuery(query= queryinput, textQuery= querytext, k=5, weightChoice= 1)
cQuery

ranking                                    title  \
15733        1  Adorable Mid Century Modern Beach Cabin   
14939        2               Charming Beach Front House   
13294        3                    Surfside paradise # 1   
18000        4                       1026 Bluewater Hwy   
15586        5                        Back To The Beach   

                                             description            city  \
15733  Newly remodeled second row beach house with ne...        Freeport   
14939  Charming &amp; colorful beach house. 3 bedroom...        Freeport   
13294  Comfortably sleeps 10 people. Open floor plan,...  Surfside Beach   
18000  Modern beach house on ocean front with direct ...  Surfside Beach   
15586  Our 3 bedroom 2 bath beach house can comfortab...  Surfside Beach   

                                                     url  
15733  https://www.airbnb.com/rooms/14185321?location...  
14939  https://www.airbnb.com/rooms/4532740?location=...  
13294  https://www.airbnb.com/rooms/15073135?location...  
18000  https://www.airbnb.com/rooms/13012994?location...  
15586  https://www.airbnb.com/rooms/18009041?location...

#### Location oriented

In [36]:
# weightChoice= 2
cQuery = airbnb.customQuery(query= queryinput, textQuery= querytext, k=5, weightChoice= 2)
cQuery

ranking                                              title  \
14862        1  Salt Grass- Christmas Bay Great Fishing & Kaya...   
17971        2                    Beautiful custom home ON beach!   
15733        3            Adorable Mid Century Modern Beach Cabin   
14939        4                         Charming Beach Front House   
13294        5                              Surfside paradise # 1   

                                             description            city  \
14862  This Bay house is close to the beach and sits ...        Freeport   
17971  Welcome to our home, Sea Dreams Beach House! W...        Freeport   
15733  Newly remodeled second row beach house with ne...        Freeport   
14939  Charming &amp; colorful beach house. 3 bedroom...        Freeport   
13294  Comfortably sleeps 10 people. Open floor plan,...  Surfside Beach   

                                                     url  
14862  https://www.airbnb.com/rooms/13186000?location...  
17971  https://www.airbnb.com/rooms/6179121?location=...  
15733  https://www.airbnb.com/rooms/14185321?location...  
14939  https://www.airbnb.com/rooms/4532740?location=...  
13294  https://www.airbnb.com/rooms/15073135?location...

#### Wrong option notification

In order to be user friendly, the user is notified when it is inputed a wrong weight option.

In [51]:
cQuery = airbnb.customQuery(query= queryinput, textQuery= querytext, k=5, weightChoice= 3)

[LOG]: You have chosen a non-existent choice
[LOG]: The possible choices are:
	[0]:= Fit-for-all
	[1]:= Price oriented
	[2]:= Location oriented
[LOG]: The [0]Fit-for-all solution has been selected by default.


In [52]:
cQuery

ranking                                              title  \
14862        1  Salt Grass- Christmas Bay Great Fishing & Kaya...   
15733        2            Adorable Mid Century Modern Beach Cabin   
14939        3                         Charming Beach Front House   
13294        4                              Surfside paradise # 1   
17971        5                    Beautiful custom home ON beach!   

                                             description            city  \
14862  This Bay house is close to the beach and sits ...        Freeport   
15733  Newly remodeled second row beach house with ne...        Freeport   
14939  Charming &amp; colorful beach house. 3 bedroom...        Freeport   
13294  Comfortably sleeps 10 people. Open floor plan,...  Surfside Beach   
17971  Welcome to our home, Sea Dreams Beach House! W...        Freeport   

                                                     url  
14862  https://www.airbnb.com/rooms/13186000?location...  
15733  https://www.airbnb.com/rooms/14185321?location...  
14939  https://www.airbnb.com/rooms/4532740?location=...  
13294  https://www.airbnb.com/rooms/15073135?location...  
17971  https://www.airbnb.com/rooms/6179121?location=...

# Bonus Step: Make a nice visualization!

This section consists in displaying the user a map with the houses available in a certain location.

The user will input a latitude and longitude of a specific location and the radius of the search. On a platform the user could click on a map and the code acquire this location.

In [3]:
location =  list(map(float, input('Choose a location: ').split(', ')))
coordinates = (float(location[0]), float(location[1]))
max_distance = input("Insert the radius distance in meters: ")
#test (32.779594, -96.800768)

#Intialize the folium map
m = folium.Map(location = coordinates, zoom_start=13)

Choose a location: 32.779594, -96.800768
Insert the radius distance in meters: 2000


Creating a circle with the radio of the search inputed by the user and adding it to the map.

In [74]:
folium.Circle(
    radius= float(max_distance),
    location=(coordinates),
    color='#3186cc',
    fill=True,
    fill_color='#3186cc'
).add_to(m)

In [ ]:
m

In this step we go through every document, retrieve the house location and price and save it in a new list.

In [79]:
dir_path = airbnb.dir_path
review_dir = airbnb.review_dir
l_temp =[]
for filePath in listdir(dir_path+review_dir):
    data = pd.read_csv(dir_path+review_dir+ filePath, delimiter = '\t',header = None, encoding = 'utf8')
    latitude = float(data.values[0][5])
    longitude = float(data.values[0][6])
    #title = str(data.values[0][7]).strip()
    average_rate_per_night = str(data.values[0][0]).strip()
    l_temp.append((latitude, longitude, average_rate_per_night))    

Now that we have the information of every house available, let's see which of those have a distance to the target location smaller than the radius inputed by the user and mark it on the map.

In [81]:
folium.Marker(location = coordinates,
              icon=folium.Icon(color='green',icon='home')
             ).add_to(m)
for i in l_temp: 
    if (distance.geodesic(coordinates, (i[0],i[1]), ellipsoid = (6377., 6356., 1 / 297.)).meters <= float(max_distance)):
        folium.Marker(location = (i[0],i[1]), popup = i[2], icon = folium.Icon(color ='blue', icon='home')
                     ).add_to(m)

Saving the results

In [83]:
m.save(os.path.join('results', r"C:\Users\guilh\Desktop\ADM\HW3\HW03\map.html"))

The created is available on this repository. Click [here](https://github.com/guilhermevescovi/Airbnb-Search-Engine/blob/master/map.html) to go to it.

In [ ]:
m